reps/2023-03-10-batch-inference.md

```diff
@@ -0,0 +1,150 @@
+
+### General Motivation
+There are changes to Ray Datasets in 2.3/2.4:
+1. Ray Datasets became lazy by default in Ray 2.3. Execution is no longer triggered until consumption is explicitly called.
+2. In Ray 2.4, Ray Datasets will move to streaming execution by default, resulting in there no longer being a separation between `Dataset` and `DatasetPipeline`. 
+3. In Ray 2.4, Ray Datasets will support a `"zero_copy"` batch format, allowing the core Ray Data to decide the most optimal batch format to use.
+
+With these changes, certain API choices that were made for the AIR batch prediction API are no longer the most optimal for simplicity and usability.
+1. Previously, AIR would manually fuse preprocessing and prediction stages, requiring `BatchPredictor` to be aware of the preprocessors passed in via a `Checkpoint`. This is not needed anymore with Datasets lazy by default, as the fusion logic can be pushed down to the Datasets layer.
+2. Previously, the separation of `Dataset` and `DatasetPipeline` resulted in `BatchPredictor` needing 2 analgous APIs: `predict` and `predict_pipelined`. This is not needed anymore with streaming as default execution.
+3. Previously, the logic for determining batch format was decided in the AIR layer in `Preprocessor` and `BatchPredictor`. This is not needed anymore with zero-copy batch format, as this decision making can be pushed down to the Datasets layer.
+
+### Should this change be within `ray` or outside?
+main `ray` project. Changes are made to Ray Data and Ray AIR level.
+
+## Stewardship
+### Required Reviewers
+The proposal will be open to the public, but please suggest a few experience Ray contributors in this technical domain whose comments will help this proposal. Ideally, the list should include Ray committers.
+
+@ericl, @clarkzinzow, @c21, @gjoliver
+
+### Shepherd of the Proposal (should be a senior committer)
+To make the review process more productive, the owner of each proposal should identify a **shepherd** (should be a senior Ray committer). The shepherd is responsible for working with the owner and making sure the proposal is in good shape (with necessary information) before marking it as ready for broader review.
+
+@ericl
+
+## Design and Architecture
+
+## API Changes
+
+1. Introduce `map_preprocessor` API directly to Ray Datasets.
+
+```python
+def map_preprocessor(preprocessor: Preprocessor, batch_size: Optional[int], prefetch_batches: Optional[int]):
+  """Apply the transform of the provided preprocessor to batches of data.
+
+  Args:
+    preprocessor: A :class:`~ray.data.preprocessor.Preprocessor` used to transform batches of data. The provided preprocessor is
+      expected to already be fitted if it is fittable.
+    batch_size: Same as `map_batches`
+    prefetch_batches: Same as `map_batches`
+  """
+```
+
+2. Introduce a `map_predictor` API directly to Ray Datasets.
+
+```python
+def map_predictor(predictor: Union[Predictor, Callable[[], Predictor]], 
```

Comment:
Maybe it should always be a lambda. I'm not sure if most predictors are copyable.

Reply:
updated